lib/solve.js

```diff
@@ -33,25 +39,126 @@ api.solve = async ({
+      engine: _engineOpts, input, graphSize, edgeCount, nonce,
+      maxNonces: noncesRemaining
+    });
+    //console.log('SOL', result);
+    /*
+    console.log('SOL', {
+      nonces: result.solutions.map(s => s.nonce),
+      time: result.time
+    });
+    */
+    if(result.solutions.length === 0) {
+      break;
+    }
+    for(const solution of result.solutions) {
+      //console.log('SOLDIFF', solution.nonce, checkDifficulty({edges: solution.edges, difficulty}));
+      if(checkDifficulty({edges: solution.edges, difficulty})) {
+        return solution;
+      }
+    }
+    const nextNonce = Math.max(...result.solutions.map(s => s.nonce)) + 1;
+    noncesRemaining -= (nextNonce - nonce);
+    nonce = nextNonce;
+  }
+  // no more nonces to check: solution not found
+  return null;
 };
+
+/**
+ * Attempts to solve the Cuckoo Cycle Proof-of-Work for the given parameters.
+ * This variation chains Cuckoo proofs together.
+ *
+ * TODO: Support Uint8Arrays for use in browser?
+ *
+ * @param engine the name of the engine to use, or an object with a name
+ *          property of the engine and other custom engine options.
+ * @param input a Node.js Buffer of up to 32 bytes that the solution is
+ *          based on, usually the output of a hash function, like SHA-256, on
+ *          some data that the proof is bound to.
+ * @param [graphSize] the size of the graph.
+ * @param [edgeCount] the expected number of edges in the solution.
+ * @param [difficulty] a 32 byte Node.js Buffer representing a target difficulty
+ *          (0 < T < 2^256) that the blake2b hash of the sorted edgelist
+ *          (ascending and serialized little endian) must be less than.
+ * @param [maxNonces] maximum number of nonces to check for all links.
+ * @param [links] number of chain links to create.
+ *
+ * @return a Promise that resolves to an array of object solutions with:
+ *          nonce the nonce as an integer.
+ *          edges an array of integers representing the edge list.
+ */
+api.solveChain = async ({
+  engine,
+  input,
+  // graph size is measured by `N` where graph is 2^N nodes
+  graphSize = constants.DEFAULT_GRAPH_SIZE,
+  edgeCount = constants.DEFAULT_EDGE_COUNT,
+  difficulty = constants.DEFAULT_DIFFICULTY,
+  maxNonces = 0xFFFFFFFF,
+  links = 1
+}) => {
+  if(!(Number.isInteger(links) && links >= 1)) {
+    throw new TypeError('"links" must be an integer > 1.');
+  }
+  const solution = [];
+  let nextInput = input;
+  for(let i = 0; i < links; ++i) {
+    const s = await api.solve({
+      engine, input: nextInput, graphSize, edgeCount, difficulty, nonce: 0,
+      maxNonces
+    });
+    if(s === null) {
+      throw new Error('Chained solution not found.');
+    }
+    solution.push(s);
+    // build next input buffer from previous solution
+    // FIXME: document format
+    const hash = crypto.createHash('sha256');
+    hash.update(Int32Array.from(s.nonce));
+    hash.update(Int32Array.from(s.edges));
```

Comment:
My understanding is that `Int32Array.from` does endianness based on the underlying architecture ... which won't work here. We need to make sure it's always little endian.

Reply:
Something like this might work as a helper function:

```js
function toUint32Array(x) {
  const b = Buffer.allocUnsafe(x.length * 4);
  x.forEach((v, i) => b.writeUInt32LE(v, i * 4));
  return b;
}
```